### DISCUSSION TITLE
Disk going to read-only mode, caused a lot of traffic to Netdata Cloud

### DISCUSSION THREAD

**Discussion by cakrit** [netdata/netdata#12857] [Q&A]
Got this from one of our best users:

> Just letting you know we had an issue today where one of our systems storage went into read-only mode and then netdata started using lots of cellular data talking to netdata cloud, it was using 1GB per hour for about 7 hours, it settled down once we rebooted the computer and the SSD was back in read/write mode. No logs as the SSD was in read-only mode.

It's obvious that we're going into a bad state when a disk is no longer writeable. Perhaps we can do better? I haven't investigated what exactly happens in this scenario. Maybe @underhood you can take a look and perhaps we can come up with ideas to address this?

It's not just the cellular data, Netdata could potentially handle such cases very gracefully and continue operating normally without writing anything to disk (even if collected data is lost at some point, it's ok).

Haven't opened it a bug, as I'm not sure if we ever supported such cases properly.

**Comment by MrZammler** [comment]
One guess would be that it failed to mark chart/alert messages in sqlite as sent, thus re-trying and re-sending them again and again... Similar case with the disk going full, we need to detect both. Not sure how easy it would be to keep everything running after this, maybe put it in a safe-mode or similar...

**Reply by cakrit** [reply]
It could perhaps switch to memory mode ram temporarily?

**Reply by underhood** [reply]
hmm... perhaps notifying cloud about it and stopping streaming would be more reasonable :/ also does sqlite3 give us reasonable feedback when that happens (to check)?

**Reply by stelfrag** [reply]
We should detect this probably in a few places and do not attempt to write anything e.g. in the database since it will fail (also trying to send anything to the cloud in this case will fail since it will need to update the status in the database -- this may be the case of having those messages not marked properly as submitted and causing constant re-trasmission) 

With RAM more we don't have an issue, with dbengine we can probably make it so that it will switch to "memory mode" internally (it will not be able to flush pages to disk which in in case it should understand that it is "ok" since we don't have a writable disk)

**Reply by Ferroin** [reply]
IMO, we probably should be improving handling of write errors in general in the dbengine.

The three big cases we could handle better here IMO are:

- A run of EIOs within a fixed window should stop writes until the agent gets restarted. If we see more than about 5% of our writes over the course of a few minutes returning this, the storage device is either broken or dying and we can not reasonably expect this to be resolved without the system being restarted.
- ENOSPC should result in some kind of backoff on writes (ideally exponential). This is something that _may_ be resolved by user intervention while the agent is still running, but us repeatedly trying to flush data to disk will actually make it more difficult to resolve.
- ENOBUFS is a similar case to ENOSPC, backoff and retry later.
